src/editor/particle_editor.cpp

```diff
@@ -0,0 +1,813 @@
+        std::string dirname = FileSystem::dirname(filepath);
+        if (!PHYSFS_mkdir(dirname.c_str()))
+        {
+          std::ostringstream msg;
+          msg << "Couldn't create directory for particle config '"
+              << dirname << "': " <<PHYSFS_getLastErrorCode();
+          throw std::runtime_error(msg.str());
+        }
+      }
+      save(filepath, true);
+    }
+  }
+}
+
+void
+ParticleEditor::save(Writer& writer)
+{
+  writer.start_list("supertux-custom-particle");
+  // Starts writing to supertux particle file. Keep this at the very beginning.
+
+  m_particles->save(writer);
+
+  // Ends writing to supertux particle file. Keep this at the very end.
+  writer.end_list("supertux-custom-particle");
+}
+
+void
+ParticleEditor::request_save(bool is_save_as, std::function<void(bool)> callback)
+{
+  if (is_save_as || m_filename == "")
+  {
+    m_enabled = false;
+    MenuManager::instance().set_menu(std::make_unique<ParticleEditorSaveAs>(callback));
+  }
+  else
+  {
+    save(m_filename);
+    callback(true);
+  }
+}
+
+void
+ParticleEditor::draw(Compositor& compositor)
+{
+  auto& context = compositor.make_context();
+
+  m_particles->draw(context);
+
+  /*context.color().draw_filled_rect(Rectf(0.f, 0.f, 255.f, context.get_height()),
+                                   Color(),
+                                   LAYER_GUI - 1);*/
+  context.color().draw_gradient(Color(0.05f, 0.1f, 0.1f, 1.f),
+                                Color(0.1f, 0.15f, 0.15f, 1.f),
+                                LAYER_GUI - 1,
+                                GradientDirection::HORIZONTAL,
+                                Rectf(0.f, 0.f, 355.f, float(context.get_height())));
+
+  if (m_in_texture_tab)
+  {
+    context.color().draw_surface_scaled((m_particles->m_textures.begin() + m_texture_current)->texture,
+                                         Rect(75, 150, 275, 350), LAYER_GUI);
+    for(const auto& control : m_controls_textures)
+    {
+      control->draw(context);
+    }
+  }
+  else
+  {
+    for(const auto& control : m_controls)
+    {
+      control->draw(context);
+    }
+  }
+
+  MouseCursor::current()->draw(context);
+}
+
+void
+ParticleEditor::update(float dt_sec, const Controller& controller)
+{
+
+  update_keyboard(controller);
+
+  if (m_quit_request) {
+    quit_editor();
+  }
+
+  if (m_in_texture_tab) {
+    for(const auto& control : m_controls_textures) {
+      control->update(dt_sec, controller);
+    }
+  } else {
+    for(const auto& control : m_controls) {
+      control->update(dt_sec, controller);
+    }
+  }
+
+  // Uncomment to make the particles stop updating on pause
+  //if (!m_enabled)
+  //  return;
+
+  m_particles->update(dt_sec);
+}
+
+void
+ParticleEditor::update_keyboard(const Controller& controller)
+{
+  if (!m_enabled)
+    return;
+
+  if (controller.pressed(Control::ESCAPE)) {
+    m_enabled = false;
+    MenuManager::instance().set_menu(MenuStorage::PARTICLE_EDITOR_MENU);
+    return;
```

Comment:
No need for this return since we're at the end of the function anyway.

Reply:
I left the `return` there in case someone might want to add more keyboard binds, so that they won't have to think about adding a `return` before adding their stuff.

Should I remove it anyways?